KEP-2008: Provide details about additional checkpoint/restore use cases

adrianreber opened the pull request:
- One-line PR description: Provide details about additional checkpoint/restore use cases

- Issue link: https://github.com/kubernetes/enhancements/issues/2008

bart0sh commented:
/ok-to-test

mikebrow reviewed (commented):
additional future stories read well...

Did you want to add more details to the forensics concept? Including a scenario how/why kubelet needs to provide garbage collection / manage which are collected..

adrianreber commented:
@mikebrow Thanks

> additional future stories read well... not obvious if they are to be officially supported cases or additional will not be supported cases

Added a paragraph that future stories currently already can be used with the existing implementation, but that additional support for a more user friendly implementations can be implemented.  Not very concrete, but acknowledged that additional support can be added.

> Did you want to add more details to the forensics story? Including a scenario how/why kubelet needs to provide garbage collection / manage which are garbage collected..

I added a paragraph about checkpoint archive management. That is might be different depending on the use case but I gave three examples how it could be implemented.

> Similarly, for cli enablement [kubernetes/kubernetes#120898 (comment)](https://github.com/kubernetes/kubernetes/pull/120898#issuecomment-1773497435) do you want to go over the expansion of this kubelet endpoint to cover kubectl enablement, mention how a user would use kubectl in these stories.. and probably add the management (manual and automatic .. CRUD) options?

I mentioned my kubectl PR in the Future Enhancements section.

PTAL.

rst0git commented on keps/sig-node/2008-forensic-container-checkpointing/README.md:
```diff
@@ -140,10 +168,93 @@ continue to run without knowing a copy was created. This copy can then
+
+#### Container Migration
+
+On of the main use cases for checkpointing and restoring containers is
```
```Suggestion
One of the main use cases for checkpointing and restoring containers is
```

rst0git commented on keps/sig-node/2008-forensic-container-checkpointing/README.md:
```diff
@@ -140,10 +168,93 @@ continue to run without knowing a copy was created. This copy can then
+
+This might be especially interesting for AI training containers which can be
+preempted for higher priority tasks without having to restart the training
+from the start.
```
Should we also mention that this is currently used with Singularity (https://arxiv.org/pdf/2202.07848.pdf)?

rst0git commented on keps/sig-node/2008-forensic-container-checkpointing/README.md:
```diff
@@ -140,10 +168,93 @@ continue to run without knowing a copy was created. This copy can then
+Alpha][kubernetes-blog-post] for details). If all these steps could automatically
+be performed by the scheduler this would greatly improve the user experience.
+Scheduler integration, however, is probably one use case which will be
+implemented at the very end and something for future enhancements.
```
It might be worth noting that the scheduler could keep track of available resources in heterogenous clusters and automatically identify suitable nodes where a checkpoint can be restored (e.g., GPU nodes with MI100 vs MI210).

adrianreber commented:
@rst0git Thanks for the suggestions and the patch. Applied.

adrianreber commented:
@mikebrow do you have any additional comments? Any recommendations how to move this forward?

suhuruli commented on keps/sig-node/2008-forensic-container-checkpointing/README.md:
```diff
@@ -132,6 +142,24 @@ message CheckpointContainerResponse {}
 
 ### User Stories
```
@adrianreber - can we add a user story around leveraging it to save costs of using GPUs (for things like training models)? This could go under Container Migration or just be an addendum to fault tolerance. 

Something along the lines of: 

With GPUs becoming a costly commodity, there is an opportunity to help customers save on costs by leveraging container checkpointing to prevent re-computation if there are any faults.

adrianreber replied:
Added. Thanks.

soltysh commented:
@adrianreber any particular reason this is separate from https://github.com/kubernetes/enhancements/pull/4288, can we get both PRs merged as one?

adrianreber commented:
> @adrianreber any particular reason this is separate from #4288, can we get both PRs merged as one?

This is again complicated. My initial approach to get this feature in Kubernetes around 2019 was to focus on container migration. Then we came up with a much simpler way using the "Forensic Container Checkpointing" story. So we worked on that and it got merged. One of the steps to get it from Alpha to Beta at that time was at least one container engine should implement the new CRI API calls. CRI-O did.

When trying to graduate from Alpha to Beta the feedback from SIG-Node was that containerd should also implement it. From containerd's side I was told that the current KEP focussing on "Forensic Container Checkpointing" feels incomplete and all possible use cases should be described in the KEP.

So now I have a Alpha to Beta graduation KEP which depends on containerd support. containerd wants to merge the `CheckpointContainer` CRI implementation PR after this one here with the complete story is merged.

The Alpha to Beta graduation is full of discussions and adding additional discussions about use cases seems to complicate everything even more.

If everybody is happy with the changes of this PR I can include it in the other PR. I don't know.

soltysh commented:
I'll defer to sig-node leads about that decision, it's theirs call about merging, I'm asking only from PRR review perspective.

adrianreber commented:
/remove-lifecycle stale
